src/Microsoft.AspNet.OData/Routing/ODataActionSelector.cs

```diff
@@ -106,5 +112,32 @@ private static HttpResponseMessage CreateErrorResponse(HttpRequestMessage reques
 
             return request.CreateErrorResponse(statusCode, error);
         }
+
+        private static bool ActionParametersMatchRequest(HttpActionDescriptor action, HttpControllerContext context)
+        {
+            var parameters = action.GetParameters();
+            var routeData = context.RouteData;
+            var routingConventionsStore = context.Request.ODataProperties().RoutingConventionsStore;
+
+            int keyCount = 0;
+            if (routingConventionsStore.ContainsKey(ODataRouteConstants.KeyCount))
+            {
+                keyCount = (int)routingConventionsStore[ODataRouteConstants.KeyCount];
+            }
+
+            // navigationProperty is optional in some cases, therefore an action
+            // should not be rejected simply because it does not declare a navigationProperty parameter
+            if (routeData.Values.ContainsKey(ODataRouteConstants.NavigationProperty))
+            {
+                keyCount -= 1;
+            }
+
+            if (parameters.Count < keyCount)
```

Comment:
Or is that only a likelyhood when there's a navigation property?

Reply:
In this case it will return `true`  if it reaches this point. If `parameters.Count` is less than the number of keys, it means that there are not enough parameters on the action to match the keys in the uri. But if `parameters.Count` is greater, it could mean that there are optional parameters on the action or parameters that are bound by values other than those in the uri.

Something else to note is that the `ODataActionSelector` in Asp.Net classic works a bit differently. It uses Asp.Net's internal action selector to return the correct match, so the method I added here just tries to ensure that the selected match is not the wrong one, that's why it does not implement rules as strict as the one in AspNetCore